src/contexts/TodosContext.tsx

```diff
@@ -0,0 +1,287 @@
+    }
+
+    const tempTodoData: Todo = {
+      title: normalizedTitle,
+      userId: USER_ID,
+      completed: false,
+      id: 0,
+      createdAt: '',
+      updatedAt: '',
+    };
+    const addTodoPromise = addTodo(tempTodoData);
+
+    try {
+      setTempTodo(tempTodoData);
+      setTodosIdsUpdating([tempTodoData.id]);
+
+      const response = await addTodoPromise;
+      const updatedTodo = { ...tempTodoData, id: response.id };
+
+      setTodos(prevTodos => [...prevTodos, updatedTodo]);
+
+      return true;
+    } catch {
+      setError(UNABLE_ADD_ERROR_MESSAGE);
+
+      return false;
+    } finally {
+      setIsInputFocused(true);
+      setTempTodo(null);
+      setTodosIdsUpdating([]);
+    }
+  };
+
+  const handleChangeTodo = async (id: number, data: TodoUpdateData) => {
+    setErrorMessage(null);
+    setTodosIdsUpdating([id]);
+
+    try {
+      const updatedTodo = await updateTodo(id, data);
+
+      setTodos(currentTodos => {
+        const index = currentTodos.findIndex(todo => todo.id === id);
+        const updatedTodos = [...currentTodos];
+
+        updatedTodos.splice(index, 1, updatedTodo);
+
+        return updatedTodos;
+      });
+    } catch {
+      setError(UNABLE_UPDATE_ERROR_MESSAGE);
+      throw new Error();
+    } finally {
+      setIsInputFocused(false);
+      setTodosIdsUpdating([]);
+    }
+  };
+
+  const handleToggleTodo = async () => {
+    const todosToChange = todos.some(todo => !todo.completed)
+      ? filterBy(todos, FilterType.Active)
+      : todos;
+    const todosId = todosToChange.map(todo => todo.id);
+
+    setTodosIdsUpdating(todosId);
+
+    const changePromises = todosToChange
+      .map(async (todo) => {
+        try {
+          return await updateTodo(todo.id, { completed: !todo.completed });
+        } catch {
+          return null;
+        }
+      });
+
+    const changedTodos = await Promise.all(changePromises);
+
+    if (changedTodos.includes(null)) {
+      setError(UNABLE_UPDATE_ERROR_MESSAGE);
+    }
+
+    setTodos((prevState) => {
+      return prevState.map((todo) => {
+        const changed = changedTodos
+          .find((changedTodo) => {
+            if (!changedTodo) {
+              return false;
+            }
+
+            return changedTodo.id === todo.id;
+          });
+
+        if (changed) {
+          return changed;
+        }
+
+        return todo;
+      });
+    });
+
+    setIsInputFocused(false);
+    setTodosIdsUpdating([]);
+  };
+
+  const handleDeleteTodo = async (id: number) => {
+    setTodosIdsUpdating([id]);
+
+    try {
+      await deleteTodo(id);
+      setTodos((prevState) => prevState.filter((todo) => todo.id !== id));
+    } catch {
+      setError(UNABLE_DELETE_ERROR_MESSAGE);
+    } finally {
+      setTodosIdsUpdating([]);
+    }
+  };
+
+  const handleDeleteCompletedTodo = async () => {
+    setErrorMessage(null);
+    const completedTodosId = todos
+      .filter(todo => todo.completed)
+      .map(todo => todo.id);
```

Comment:
Use the `reduce` method to walk the array once instead of twice and improve the performance of your function.